### DISCUSSION TITLE
GPU selection and benchmarks

### DISCUSSION THREAD

**Discussion by quantumdot** [talmolab/sleap#959] [Help!] [answered]
Looking for information on what GPUs are recommended for use with SLEAP, particularly in a real-time or closed-loop setting. Are there benchmarks available for SLEAP on different GPUs, and possibly different model configurations?

**Comment by talmo** [comment, marked as answer]
Hi @quantumdot,

**GPUs**:
We've done most of our testing on an Titan RTX, but that's way overkill, especially for realtime. Anything with >8 GB of memory should be enough, especially for inference. Any of the 30xx series should have enough CUDA cores to get your realtime latency pretty low.

**Benchmarking**:
- [Our paper](https://www.nature.com/articles/s41592-022-01426-1#:~:text=Fast%2C%20efficient%20and%20accurate%20animal%20pose%20estimation) has a good bit of results on performance.
- The Methods has a [section on speed benchmarking](https://www.nature.com/articles/s41592-022-01426-1#Sec10:~:text=required%20after%20tracking.-,Speed%20benchmarking,-To%20evaluate%20model) which is helpful as well.

**Closed loop**:
- [DeepLabStream](https://github.com/SchwarzNeuroconLab/DeepLabStream) has native support for using SLEAP models for realtime/closed-loop experiments.
- Here's a [module for realtime inference using SLEAP](https://github.com/talmo/leap_rigs/blob/main/leap_rigs/tracking.py) that we used for our closed-loop experiments in the Nature Methods (2022) paper.
- Here's a [notebook on realtime inference using SLEAP](https://sleap.ai/notebooks/Interactive_and_realtime_inference.html) that's in our docs page (scroll to the last few sections).
- We recently added the [sleap-export](https://sleap.ai/guides/cli.html#sleap-export) command which generates TensorFlow `SavedModel`s and frozen graphs, which are compatible for use with C++ TensorFlow inference engines and should be convertible to [TensorRT](https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html) for massive performance gains (especially with FLOAT16 quantization).
- We'll soon be releasing a guide on using the new [Bonsai-RX SLEAP package](https://github.com/bonsai-rx/sleap) which gives you the ability to use SLEAP within a more sophisticated experiment control framework.


In terms of model configurations: that's tricky since it depends heavily on your dataset and accuracy requirements. The rule of thumb is that you want to first get to a model configuration that is accurate enough for your needs, and then from there start cutting down on the number of computations. A few good hyperparameters to tune for performance:
- _Approach_: Top-down will almost always be faster than any other approach since it crops a bounding box around the animal. This may even be true in single-animal settings if the animal is small relative to the size of the frame.
- _Filters and filter rate_: Decreasing these will decrease the overall number of parameters in the model, resulting in fewer computations at the expense of reduced capacity.
- _Output stride_: Increasing this to 4 or 8 can result in significant speed-ups at the cost of resolution/accuracy.


Let us know if you try any of these and have any questions.

Cheers,

Talmo

**Reply by quantumdot** [reply]
Hi Talmo,

Wow, thanks for the great, detailed answer!

Do you have any experience with professional cards, such as the RTX A2000 through the RTX A5000, or Tesla architectures?

It would be a great thing to have a benchmark script which users could run with a common set of models on a common set of videos and produce sharable benchmark statistics.

**Reply by jmdelahanty** [reply]
Our lab has several machines with (I'm pretty sure...) A4000 GPUs on board. I'd be happy to try running SLEAP on one to see how it does. I also did some stuff with messing around in benchmarking (maybe not so professionally/best practices) for our lab [here](https://github.com/talmolab/sleap/discussions/833) with related scripts [here](https://github.com/Tyelab/sleap_benchmarking/tree/main).
